portia_server/portia_api/resources/projects.py

```diff
@@ -1,29 +1,41 @@
-from collections import OrderedDict
+from collections import OrderedDict, defaultdict
+from itertools import chain
 
 from django.utils.functional import cached_property
 from rest_framework.decorators import detail_route
 from rest_framework.response import Response
-from rest_framework.status import HTTP_200_OK
+from rest_framework.status import HTTP_200_OK, HTTP_201_CREATED
 from six import iteritems
 
 from portia_orm.models import Project
 from storage import get_storage_class
 from .route import (JsonApiRoute, JsonApiModelRoute,
                     ListModelMixin, RetrieveModelMixin)
-from ..jsonapi.exceptions import JsonApiFeatureNotAvailableError
+from .response import FileResponse
+from ..jsonapi.exceptions import (JsonApiFeatureNotAvailableError,
+                                  JsonApiBadRequestError,
+                                  JsonApiNotFoundError,
+                                  JsonApiConflictError)
+from ..utils.download import ProjectArchiver, CodeProjectArchiver
 
 
 class ProjectDownloadMixin(object):
     @detail_route(methods=['get'])
     def download(self, *args, **kwargs):
-        # project_manager = self.project_manager
-        project_id = self.kwargs.get('project_id')
-        spider_id = self.kwargs.get('spider_id', None)
-        spider_ids = [spider_id] if spider_id is not None else '*'
         fmt = self.query.get('format', 'spec')
-        # return ProjectDownloadResponse(
-        #     project_id, spider_ids, fmt, project_manager)
-        return Response(b'', status=HTTP_200_OK)
+        version = self.query.get('version', None)
+        branch = self.query.get('branch', None)
+        spider_id = self.kwargs.get('spider_id', None)
+        spiders = [spider_id] if spider_id is not None else None
+        if self.storage.version_control and (version or branch):
+            storage = self.storage.__class__(
+                self.storage.name, self.storage.author, version, branch)
```

Comment:
We can make this a constructor method as an instance method which receives a version and branch.